### DISCUSSION TITLE
Design for xml writer

### DISCUSSION THREAD

**Discussion by AdRiley** [enso-org/enso#8872] [Work In Progress - Libraries]
Some initial design ideas for a simple XML writer.

These mock ups show a to_xml method on Table, but the same API could be applied to a file writer with some additional design.

to_xml takes 5 parameters (all optional)

- attribute_columns = []
- element_columns = []
- root_element_name="root"
- row_element_name="row"
- error_on_missing_columns=True

In the default case where nothing is set we get an XML document where each row of data is an xml element called row and all columns become attributes.

![image](https://github.com/enso-org/enso/assets/1720119/46600552-afbb-4c75-816a-8e0895a97e55)

If you specify a column to be an attribute, all unspecified columns become elements

![image](https://github.com/enso-org/enso/assets/1720119/9cae63dc-6735-4b64-bfb9-1febfe43954a)

Likewise if you specify a column to be an attribute, all unspecified columns become attributes

![image](https://github.com/enso-org/enso/assets/1720119/cf6d7225-b20e-44df-9778-34193efa91eb)

If you specify attributes and elements then you will only get those specified columns in the xml. Plus a warning that you have fields you are not using. (The warning can be hushed via the error_on_missing_columns parameter)

![image](https://github.com/enso-org/enso/assets/1720119/7fc63438-baed-4340-93ba-35cd57f25e68)

The root_element_name and row_element_name parameters allow customising of the respective element names

![image](https://github.com/enso-org/enso/assets/1720119/404f7f2c-f2a0-480a-a785-8e89ead6f913)

**Comment by jdunkerley** [comment]
I'd prefer defaulting to being child nodes rather than attributes and liked the idea of using `@` to allow controlling this just from column names.

We should have the option of a "value column" as well. So you could have attributes and text as an output.

What will we do if a column is invalid XML node name - normalise?

**Reply by AdRiley** [reply]
I don't like the idea of "magic" column names.
Agree value column is needed.
Yes we will normalise names to valid xml.

**Reply by AdRiley** [reply]
Don't mind on what the default is so happy to switch it to child nodes.

**Comment by radeusgd** [comment]
> (The warning can be hushed via the error_on_missing_columns parameter)

I'm not sure if this is right to control a _warning_ through a parameter called `error_*`. Wouldn't the _'standard'_ `on_problems` parameter suit here?

**Reply by AdRiley** [reply]
Yes you are right. That would be better.

**Comment by GregoryTravis** [comment]
There are two behaviors here that are implicitly set by the values of `attribute_columns` and `element_columns`:
* Should a column be included in the output?
* If a column is included, is it an attribute or an element?

A column is included if it is explicitly mentioned, or if only one (but not both) of `attribute_columns` and `element_columns` are expclitily set.
A column is assigned to be an attribute or element (1) if it is explicitly included in one of the two column lists, or implicitly if it is not mentioned in column lists *and only one of the column list parameters is used*.

I worry that this might be tricky to understand.

A simpler model might be: a column is only included if it is specified in one of the two column lists. Also, we default `attribute_columns` to contain all fields, giving the user a way to immediately see why columns are included and in what way. (But this would require that the default value be visible in the node immediately on creation, which I don't think we support.)

**Reply by radeusgd** [reply]
+1 to that approach.

What I don't like about 
> If you specify a column to be an attribute, all unspecified columns become elements
is that - what if I want to just keep that column as attribute and ignore other columns altogether?

Well of course I can just drop (`remove_columns`) them in an earlier step.

But it feel strange that setting one column to be attribute may automatically 'flip' all the other columns to be elements instead.

I like the more explicit approach that @GregoryTravis is suggesting - it feels more predictable

**Reply by AdRiley** [reply]
Yes maybe the default behaviour is trying to be too clever. Odds are if you are creating XML you are doing it to an existing specification so are going to need to configure the component.

**Reply by AdRiley** [reply]
I am convinced by the more explicit (and less clever) approach.

**Comment by AdRiley** [comment]
Latest API design

![image](https://github.com/enso-org/enso/assets/1720119/6361a0d7-43d6-4710-bca6-dc53f6411613)

element_columns defaults to all columns as per @GregoryTravis's suggestion.

I feel like there was feedback on what the parameter names should be, but I can't see it now.

**Reply by AdRiley** [reply]
Any objections if I remove error_on_missing_columns? And make the component always error in that case.
